### PR TITLE
PET Scatter

### PR DESCRIPTION
**Pull request by NikEfth**
This a different approach by sub-classing.

### CODE REVIEW

**Comment by KrisThielemans** [comment]
can someone update the Travsis.yml to point to the correct STIR branch? @rijobro @casperdcl @paskino ?

**Comment by casperdcl** [comment]
done I think

File: .travis.yml [hunk 1]
```diff
@@ -68,7 +68,7 @@ matrix:
 env:
  global:
   # Need STIR_TAG=master at present, see https://github.com/CCPPETMR/SIRF/issues/366
-  - BUILD_FLAGS="-DCMAKE_BUILD_TYPE=Release -DSTIR_TAG=master"
+  - BUILD_FLAGS="-DCMAKE_BUILD_TYPE=Release -DSTIR_URL=https://github.com/NikEfth/STIR -DSTIR_TAG=Scatter_simulation -DSTIR_BUILD_EXECUTABLES=ON"
```
Comment (KrisThielemans) [inline .travis.yml]:
```suggestion
  - BUILD_FLAGS="-DCMAKE_BUILD_TYPE=Release -DSTIR_URL=https://github.com/NikEfth/STIR -DSTIR_TAG=Scatter_simulation"
```
@casperdcl we shouldn't need it for SIRF (I put it in the email to be able to run the STIR recon_test_pack). let's first see if it gets through.

File: examples/Python/PET/scatter_simulation.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    sss.set_acquisition_data(tmpl_acq_data)
+
+    # Set the attenuation image
+    sss.set_attenuation_image(atten_image)
+
+    # Set the activity image
+    sss.set_activity_image(act_image)
+
+    # Run the scatter simulation
+    sss.run_scatter_simulation()
```
Comment (KrisThielemans) [inline examples/Python/PET/scatter_simulation.py]: call this `process()` like other SIRF classes

File: examples/Python/PET/scatter_simulation.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+    # Get the simulated sinogram.
+    # Note: The sinogram has been downscaled.
+    sss_data = sss.get_simulated_data()
```
Comment (KrisThielemans) [inline examples/Python/PET/scatter_simulation.py]: call this `get_output()` like other SIRF classes

**Comment by NikEfth** [comment]
@KrisThielemans I finished (probably should say started) the minimal interface to ```ScatterEstimation```. 
It can run the estimation calling a par file, which sets everything. 
Gradually, the setters and getter will appear, but I think this is should allow Christoph or any other SIRF user to run it,  without fiddling with bash commands. 
**** 
In the example I try to call the STIR path environment path. I do not know how to do this in a proper pyhonic way. If somebody could make it look nice I would kindly appreciate it.

**Comment by evgueni-ovtchinnikov** [comment]
Tried on VM: `scatter_simulation.py` runs not reporting any errors (which of course does not necessarily mean everything is ok), but `scatter_estimation.py` throws
```
ERROR: Unknown originating_system '', when parsing file './NEMA_IQ/20170809_NEMA_UCL.n.hdr'
```
though in `./NEMA_IQ/20170809_NEMA_UCL.n.hdr` i see
```
!originating system:=2008
```

**Comment by KrisThielemans** [comment]
@evgueni-ovtchinnikov could you tell us what OS you're running this on? It seems a STIR issue. Could you check the STIR/recon_test_pack/ScatterRun?

**Comment by evgueni-ovtchinnikov** [comment]
@KrisThielemans: our VM is Ubuntu 18.04 if I am not mistaken?

What do I need to check in STIR/recon_test_pack/ScatterRun?

**Comment by KrisThielemans** [comment]
I was wondering if you were runnign on Windows. If on the VM, it's weird as @ckolbPTB is as well.

I suppose maybe an easier thing to do would be to check our mMR demo recon that uses this data.

For ScatterRun, the [README](https://github.com/UCL/STIR/tree/78650140ebe2ab6bb8412ae2965400340161c06c/recon_test_pack/ScatterRun) gives some instructions. Seems like you need to copy the data in that directory (or copy the par files elsewhere) and run `scatter_estimate scatter_estimation.par`.

It's far a too complicated way to check if STIR can read the norm file, but it should work...

**Comment by NikEfth** [comment]
@evgueni-ovtchinnikov you will also need the STIR executables

**Comment by evgueni-ovtchinnikov** [comment]
@KrisThielemans: cannot find `scatter_estimate` executable nor `scatter_estimation.par`. Found `estimate_scatter.sh`, but it wants various files (`atten_image` etc.) that are not in `ScatterRun`.

I did build with `STIR_BUILD_EXECUTABLES=ON, STIR_TAG=Scatter_simulation, STIR_URL=https://github.com/NikEfth/STIR`

**Comment by ckolbPTB** [comment]
Hi,

I did:

cmake ../SIRF-SuperBuild -DDEVEL_BUILD=ON -USTIR_URL -USTIR_TAG -USIRF_URL 
-USIRF_TAG

and then:

cmake ../SIRF-SuperBuild -DSIRF_URL=https://github.com/NikEfth/SIRF/ 
-DSIRF_TAG=scatter_example_2 -DSTIR_URL=https://github.com/NikEfth/STIR/ 
-DSTIR_TAG=Scatter_simulation -DSTIR_BUILD_EXECUTABLES=ON

to get it to work.

Regards,
Christoph
--
Christoph Kolbitsch, Ph.D.
Head of Research Group
Quantitative MRI

Physikalisch-Technische Bundesanstalt
Abbestr. 2-12
10587 Berlin, Germany
phone: +49 30 3481 7761



Von:    "Evgueni Ovtchinnikov" <notifications@github.com>
An:     "CCPPETMR/SIRF" <SIRF@noreply.github.com>
Kopie:  "ckolbPTB" <christoph.kolbitsch@ptb.de>, "Mention" 
<mention@noreply.github.com>
Datum:  25.06.2019 11:47
Betreff:        Re: [CCPPETMR/SIRF] Scatter example 2 (#404)



@KrisThielemans: cannot find scatter_estimate executable nor 
scatter_estimation.par. Found estimate_scatter.sh, but it wants various 
files (atten_image etc.) that are not in ScatterRun.
I did build with STIR_BUILD_EXECUTABLES=ON, STIR_TAG=Scatter_simulation, 
STIR_URL=https://github.com/NikEfth/STIR
?

**Comment by evgueni-ovtchinnikov** [comment]
my cmake setting are the same as Christoph's

@ckolbPTB: can you run STIR scatter executables the way Kris suggested above?

**Comment by ckolbPTB** [comment]
no, I could not run any STIR executables.

I ran python ~/devel/build/sources/SIRF/examples/Python/PET/scatter_estimation.py

in the folder /home/sirf/devel/build/sources/STIR/recon_test_pack/ScatterRun/

This is also were the parameter file ScatterEstimation.par is

File: src/xSTIR/pSTIR/STIR.py [hunk 4]
```diff
@@ -1728,6 +1728,83 @@ def __del__(self):
...
+
+    def process(self):
+        print('ScatterEstimator:: Waiting the scatter estimation to finish ...')
+        self.output = AcquisitionData()
+        self.output.handle = pystir.cSTIR_runScatterEstimation(self.handle)
+        check_status(self.output.handle)
+        print('ScatterSimulator:: Estimation finished.')
+
+    def get_output(self, est_num):
+        print('ScatterEstimator:: doing here')
```
Comment (KrisThielemans) [inline src/xSTIR/pSTIR/STIR.py]: will need something

**Comment by KrisThielemans** [comment]
All OSX jobs failing due to python dependencies lagging behind python version with brew.

**Comment by KrisThielemans** [comment]
This is as good as ready, or at least as how much energy I have. It isn't incorporated in the NEMA data yet unfortunately. A volunteer?

Result of profiles on the first view of the sum of all sinograms:
image

**Comment by KrisThielemans** [comment]
will close #138

**Comment by KrisThielemans** [comment]
Relies on https://github.com/UCL/STIR/pull/744. That PR is on STIR `release_4`, so we will need to merge that on `master` first.

Once that's done, current `.travis.yml` will need to be restored again

**Comment by KrisThielemans** [comment]
Travis failures due to unrelated #854  and a bug in STIR `master`, now fixed https://github.com/UCL/STIR/commit/754bef86516e2f61e9441cf1598aec645f07254b

**Pull request merged**